src/riak_kv_sweeper_state.erl

```diff
@@ -0,0 +1,1076 @@
+        [{expired_or_missing(Now, Sweep, Participants), Sweep} ||
+         Sweep <- get_idle_sweeps(Sweeps)],
+    case ExpiredMissingSweeps of
+        [] ->
+            [];
+        _ ->
+            MostExpiredMissingSweep =
+                hd(lists:reverse(lists:keysort(1, ExpiredMissingSweeps))),
+            case MostExpiredMissingSweep of
+                %% Non of the sweeps have a expired or missing participant.
+                {{0,0}, _} -> [];
+                {_N, Sweep} -> Sweep
+            end
+    end.
+
+-spec expired_or_missing(erlang:timestamp(),
+                         #sweep{},
+                         dict()) ->
+    {MissingSum :: non_neg_integer(),
+     ExpiredSum :: non_neg_integer()}.
+expired_or_missing(Now, #sweep{results = Results}, Participants) ->
+    ResultsList = dict:to_list(Results),
+    Missing = missing(run_interval, Participants, ResultsList),
+    Expired =
+        [begin
+             RunInterval = run_interval(Mod, Participants),
+             expired(Now, TS, RunInterval)
+         end ||
+         {Mod, {TS, _Outcome}} <- ResultsList],
+    MissingSum = lists:sum(Missing),
+    ExpiredSum = lists:sum(Expired),
+    {MissingSum, ExpiredSum}.
+
+-spec expired(erlang:timestamp(),
+              erlang:timestamp(),
+              RunInterval :: non_neg_integer() | disabled) ->
+    non_neg_integer().
+expired(_Now, _TS, disabled) ->
+    0;
+expired(Now, TS, RunInterval) ->
+    case elapsed_secs(Now, TS) - RunInterval of
+        N when N < 0 ->
+            0;
+        N ->
+            N
+    end.
+
+-spec run_interval(riak_kv_sweeper:participant_module(),
+                   Participants :: dict()) ->
+    RunInterval :: non_neg_integer() | 'disabled'.
+run_interval(Mod, Participants) ->
+    case dict:find(Mod, Participants) of
+        {ok, #sweep_participant{run_interval = RunInterval}} ->
+            get_run_interval(RunInterval);
+        _ ->
+            %% Participant have been disabled since last run.
+            %% TODO: should we remove inactive results?
+            disabled
+    end.
+
+-spec elapsed_secs(erlang:timestamp(), erlang:timestamp()) ->
+    non_neg_integer().
+elapsed_secs(Now, Start) ->
+    timer:now_diff(Now, Start) div 1000000.
+
+-spec random_sweep([#sweep{}]) -> #sweep{}.
+random_sweep(Sweeps) ->
+    Index = random:uniform(length(Sweeps)),
+    lists:nth(Index, Sweeps).
+
+-spec scheduler_enabled() -> boolean().
+scheduler_enabled() ->
+    case app_helper:get_env(riak_kv, sweeper_scheduler, true) of
+        E when E == true; E == false ->
+            E;
+        _ ->
+            false
+    end.
+
+-spec sweep_window() ->
+    'always'
+    | 'never'
+    | {StartHour :: non_neg_integer(), EndHour :: non_neg_integer()}.
+sweep_window() ->
+    case application:get_env(riak_kv, sweep_window) of
+        {ok, always} ->
+            always;
+        {ok, never} ->
+            never;
+        {ok, {StartHour, EndHour}} when StartHour >= 0, StartHour =< 23,
```

Comment:
Macro created. However, it is valid to have `StartHour` > `EndHour`. This allows you to have a sweep window from, say, 10PM to 2AM. The existing sweeper code handles this correctly.